File: libnd4j/include/ops/declarable/generic/parity_ops/unsorted_segment_prod.cpp
/* ******************************************************************************
 *
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 *  See the NOTICE file distributed with this work for additional
 *  information regarding copyright ownership.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

//
// Created by george@skymind.io on 9/6/2018.
//

#include <ops/declarable/CustomOperations.h>
#include <ops/declarable/helpers/segment.h>
#if NOT_EXCLUDED(OP_unsorted_segment_prod)
namespace sd {
namespace ops {
CUSTOM_OP_IMPL(unsorted_segment_prod, 2, 1, false, 0, 0) {
  auto input = INPUT_VARIABLE(0);
  auto reshapedInput = *input;
  /*   if(!input->isVector()) {
         reshapedInput = input->reshape('c',{input->lengthOf()},false);
     }*/

  auto idxSegments = INPUT_VARIABLE(1);
  auto reshapedSegments = *idxSegments;
  if (!idxSegments->isVector() && idxSegments->rankOf() > 1) {
    std::vector<sd::LongType> shape = {idxSegments->lengthOf()};
    reshapedSegments = idxSegments->reshape('c',shape, false);
  }

  auto segmentedOutput = OUTPUT_NULLIFIED(0);
  LongType numOfClasses = block.width() == 3 ? INPUT_VARIABLE(2)->e<LongType>(0) : INT_ARG(0);
  REQUIRE_TRUE(reshapedSegments.isVector(), 0,
               "unsorted_segment_prod: segment indexes array should be a vector, but it rank is %i.",
               idxSegments->rankOf());
  REQUIRE_TRUE(reshapedSegments.lengthOf() == input->sizeAt(0), 0,
               "unsorted_segment_pod: segment indexes array length should be equal to the input first dimension, but "
               "%ld != %ld.",
               reshapedSegments.lengthOf(), input->sizeAt(0));

  LongType wrong;

  REQUIRE_TRUE(helpers::unsortedSegmentIndicesValidate(block.launchContext(), &reshapedSegments, numOfClasses, wrong),
               0, "unsorted_segment_pod: segment indices should be in range [0, %ld), but %ld != %ld", numOfClasses,
               wrong, numOfClasses);
  helpers::unsortedSegmentProdFunctor(block.launchContext(), &reshapedInput, &reshapedSegments, numOfClasses,
                                      segmentedOutput);

  return Status::OK;
}

DECLARE_SHAPE_FN(unsorted_segment_prod) {
  auto in = inputShape->at(0);
  int outRank = shape::rank(in);
  LongType* outputShape = nullptr;
  LongType numOfClasses = block.width() == 3 ? INPUT_VARIABLE(2)->e<LongType>(0) : INT_ARG(0);

  if (INPUT_VARIABLE(0)->rankOf() >= 2) {
    ALLOCATE(outputShape, block.getWorkspace(), shape::shapeInfoLength(outRank), sd::LongType);
    outputShape[0] = outRank;
    outputShape[1] = numOfClasses;
    for (LongType i = 1; i < outRank; i++) outputShape[i + 1] = shape::sizeAt(in, i);

    ShapeUtils::updateStridesAndType(outputShape, in, shape::order(in));

  } else {
    ALLOCATE(outputShape, block.getWorkspace(), shape::shapeInfoLength(1), sd::LongType);
    outputShape[0] = 1;
    outputShape[1] = numOfClasses;
    ShapeUtils::updateStridesAndType(outputShape, in, shape::order(in));
  }

  return SHAPELIST(CONSTANT(outputShape));
}
DECLARE_TYPES(unsorted_segment_prod) {
  getOpDescriptor()
      ->setAllowedOutputTypes({ALL_FLOATS, ALL_INTS})
      ->setAllowedInputTypes(0, {ALL_FLOATS, ALL_INTS})
      ->setAllowedInputTypes(1, {ALL_INDICES})
      ->setSameMode(false);
}

CUSTOM_OP_IMPL(unsorted_segment_prod_bp, 3, 2, false, 0, 1) {
  auto input = INPUT_VARIABLE(0);
  auto indices = INPUT_VARIABLE(1);
  auto eps = INPUT_VARIABLE(2);
  //            auto numOfClasses = INT_ARG(0);
  auto output = OUTPUT_NULLIFIED(0);

  LongType numOfClasses = block.width() == 4 ? INPUT_VARIABLE(3)->e<LongType>(0) : INT_ARG(0);
  REQUIRE_TRUE(indices->isVector(), 0,
               "unsorted_segment_prod_bp: segment indexes array should be a vector, but it rank is %i.",
               indices->rankOf());
  REQUIRE_TRUE(indices->lengthOf() == input->sizeAt(0), 0,
               "unsorted_segment_prod_bp: segment indexes array length should be equal to the input first dimension, "
               "but %lld != %lld.",
               indices->lengthOf(), input->sizeAt(0));

  LongType wrong = numOfClasses;

  REQUIRE_TRUE(helpers::unsortedSegmentIndicesValidate(block.launchContext(), indices, numOfClasses, wrong), 0,
               "unsorted_segment_prod_bp: segment indices should be in range [0, %lld), but %lld > %lld", numOfClasses,
               wrong, numOfClasses);

  return helpers::unsortedSegmentProdFunctorBP(block.launchContext(), input, indices, eps, numOfClasses, output);
}
DECLARE_TYPES(unsorted_segment_prod_bp) {
  getOpDescriptor()
      ->setAllowedOutputTypes(0, {ALL_FLOATS})
      ->setAllowedOutputTypes(1, {ALL_INDICES})
      ->setAllowedInputTypes(0, {ALL_FLOATS})
      ->setAllowedInputTypes(1, {ALL_INDICES})
      ->setAllowedInputTypes(2, {ALL_FLOATS, ALL_INTS})
      ->setSameMode(false);
}

DECLARE_SHAPE_FN(unsorted_segment_prod_bp) {
  auto in = inputShape->at(0);
  auto inIdx = inputShape->at(1);
  return SHAPELIST(CONSTANT(in), CONSTANT(inIdx));
}

}  // namespace ops

}  // namespace sd
#endif

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/segment/UnsortedSegmentProd.java
/*******************************************************************************
 * Copyright (c) 2015-2018 Skymind, Inc.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

package org.nd4j.linalg.api.ops.impl.transforms.segment;

import lombok.NoArgsConstructor;
import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.base.Preconditions;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.ops.DynamicCustomOp;
import org.nd4j.linalg.api.ops.impl.transforms.segment.bp.UnsortedSegmentProdBp;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/**
 * Unsorted segment product operation
 *
 * @author Alex Black
 */
@NoArgsConstructor
public class UnsortedSegmentProd extends DynamicCustomOp {

    private int numSegments;

    public UnsortedSegmentProd(SameDiff sameDiff, SDVariable data, SDVariable segmentIds, int numSegments) {
        super(null, sameDiff,  new SDVariable[] {data, segmentIds}, false);
        this.numSegments = numSegments;
        addIArgument(numSegments);
    }

    public UnsortedSegmentProd(INDArray data, INDArray segmentIds, int numSegments) {
        addInputArgument(data, segmentIds);
        addIArgument(numSegments);
    }

    @Override
    public String opName(){
        return "unsorted_segment_prod";
    }

    @Override
    public String tensorflowName() {
        return "UnsortedSegmentProd";
    }

    @Override
    public List<SDVariable> doDiff(List<SDVariable> gradients){
        return new UnsortedSegmentProdBp(sameDiff, arg(0), arg(1), gradients.get(0), numSegments).outputs();
    }

    @Override
    public List<DataType> calculateOutputDataTypes(List<DataType> inputDataTypes){
        Preconditions.checkState(inputDataTypes != null && (inputDataTypes.size() == 2 || inputDataTypes.size() == 3),
                "Expected exactly 2 input data types for %s, got %s", getClass(), inputDataTypes);
        return Collections.singletonList(inputDataTypes.get(0));
    }
}
